File: pythonlibrary/aibrd/generators/sprint_feed.py
"""Sprint task generator — converts BRD content into developer sprint tasks."""

from __future__ import annotations
from dataclasses import dataclass, field
from typing import Literal
from datetime import date

from ..llm.client import call_llm_json
from ..models.outputs import BRDContent


Priority = Literal["high", "medium", "low"]


@dataclass
class SprintTask:
    id: str
    title: str
    requirement_ids: list[str]
    acceptance_criteria: list[str]
    story_points: int
    priority: Priority
    module: str | None = None


SYSTEM = """You are a senior tech lead converting business requirements into sprint tasks.
For each business flow or feature, create a developer task with:
- A clear action title (verb-first)
- Which requirement IDs it covers
- Acceptance criteria (from AC-XXX items)
- Story point estimate (1, 2, 3, 5, 8, 13)
- Priority based on dependencies (high/medium/low)
Return JSON: { "tasks": [{ "title": "...", "requirementIds": ["BF-001"], "acceptanceCriteria": ["..."], "storyPoints": 3, "priority": "high" }] }"""


def generate_sprint_feed(content: BRDContent) -> list[SprintTask]:
    """Generate sprint tasks from BRD content."""
    brd_summary = "\n".join([
        *[f"{f.id}: {f.name} — {f.description}" for f in content.flows],
        *[f"{r.id}: {r.description}" for r in content.rules],
        *[f"{ac.id}: Given {ac.given} / When {ac.when} / Then {ac.then}" for ac in content.criteria],
    ])

    raw = call_llm_json(brd_summary, SYSTEM)
    tasks = raw.get("tasks", [])
    return [
        SprintTask(
            id=f"TASK-{str(i + 1).zfill(3)}",
            title=t["title"],
            requirement_ids=t.get("requirementIds", []),
            acceptance_criteria=t.get("acceptanceCriteria", []),
            story_points=t.get("storyPoints", 3),
            priority=t.get("priority", "medium"),
            module=content.module_slug,
        )
        for i, t in enumerate(tasks)
    ]


def format_sprint_feed(tasks: list[SprintTask], project_name: str) -> str:
    today = date.today().isoformat()
    total_pts = sum(t.story_points for t in tasks)
    lines = [
        f"# Sprint Feed — {project_name}",
        f"_Generated: {today} | {len(tasks)} tasks · {total_pts} story points_",
        "",
    ]

    for priority in ("high", "medium", "low"):
        filtered = [t for t in tasks if t.priority == priority]
        if not filtered:
            continue
        emoji = {"high": "🔴", "medium": "🟡", "low": "🟢"}[priority]
        lines += [f"## {emoji} {priority.capitalize()} Priority", ""]
        for task in filtered:
            lines.append(f"### {task.id}: {task.title}")
            lines.append(f"**Story Points:** {task.story_points}")
            if task.module:
                lines.append(f"**Module:** {task.module}")
            lines.append(f"**Traces:** {', '.join(task.requirement_ids)}")
            if task.acceptance_criteria:
                lines.append("**Acceptance Criteria:**")
                for ac in task.acceptance_criteria:
                    lines.append(f"- [ ] {ac}")
            lines.append("")

    lines += ["## Summary Table", ""]
    lines.append("| Task | Priority | Points | Requirements |")
    lines.append("|---|---|---|---|")
    for t in tasks:
        lines.append(f"| {t.id}: {t.title} | {t.priority} | {t.story_points} | {', '.join(t.requirement_ids)} |")

    return "\n".join(lines)


def format_sprint_feed_as_github_issues(tasks: list[SprintTask]) -> str:
    """Return newline-delimited JSON for GitHub Issues API."""
    import json
    results = []
    for t in tasks:
        body_lines = [
            f"**Traces:** {', '.join(t.requirement_ids)}",
            f"**Story Points:** {t.story_points}",
            "",
            "## Acceptance Criteria",
            *[f"- [ ] {ac}" for ac in t.acceptance_criteria],
        ]
        results.append(json.dumps({
            "title": f"[{t.id}] {t.title}",
            "body": "\n".join(body_lines),
            "labels": [l for l in [t.priority, t.module or "general"] if l],
        }))
    return "\n".join(results)

File: vscode-extension/src/core/generators/sprint_feed.ts
import * as vscode from 'vscode'
import { callLLMJson } from '../../llm/client'
import { BRDContent } from '../models/outputs'

export interface SprintTask {
  id: string
  title: string
  requirementIds: string[]
  acceptanceCriteria: string[]
  storyPoints: number
  priority: 'high' | 'medium' | 'low'
  module?: string
}

const SYSTEM = `You are a senior tech lead converting business requirements into sprint tasks.
For each business flow or feature, create a developer task with:
- A clear action title (verb-first)
- Which requirement IDs it covers
- Acceptance criteria (from AC-XXX items)
- Story point estimate (1, 2, 3, 5, 8, 13)
- Priority based on dependencies (high/medium/low)
Return JSON: { "tasks": [{ "title": "...", "requirementIds": ["BF-001"], "acceptanceCriteria": ["..."], "storyPoints": 3, "priority": "high" }] }`

export async function generateSprintFeed(
  content: BRDContent,
  token?: vscode.CancellationToken
): Promise<SprintTask[]> {
  const brdSummary = [
    ...content.flows.map(f => `${f.id}: ${f.name} — ${f.description}`),
    ...content.rules.map(r => `${r.id}: ${r.description}`),
    ...content.criteria.map(ac => `${ac.id}: Given ${ac.given} / When ${ac.when} / Then ${ac.then}`)
  ].join('\n')

  const raw = await callLLMJson<{ tasks: Omit<SprintTask, 'id'>[] }>(brdSummary, SYSTEM, token)
  return raw.tasks.map((t, i) => ({
    ...t,
    id: `TASK-${String(i + 1).padStart(3, '0')}`,
    module: content.moduleSlug
  }))
}

export function formatSprintFeed(tasks: SprintTask[], projectName: string): string {
  const today = new Date().toISOString().split('T')[0]
  const total = tasks.reduce((s, t) => s + t.storyPoints, 0)

  const lines = [
    `# Sprint Feed — ${projectName}`,
    `_Generated: ${today} | ${tasks.length} tasks · ${total} story points_`,
    ''
  ]

  const byPriority = ['high', 'medium', 'low'] as const
  for (const p of byPriority) {
    const filtered = tasks.filter(t => t.priority === p)
    if (!filtered.length) continue

    const emoji = p === 'high' ? '🔴' : p === 'medium' ? '🟡' : '🟢'
    lines.push(`## ${emoji} ${p.charAt(0).toUpperCase() + p.slice(1)} Priority`, '')

    for (const task of filtered) {
      lines.push(`### ${task.id}: ${task.title}`)
      lines.push(`**Story Points:** ${task.storyPoints}`)
      if (task.module) lines.push(`**Module:** ${task.module}`)
      lines.push(`**Traces:** ${task.requirementIds.join(', ')}`)
      if (task.acceptanceCriteria.length) {
        lines.push('**Acceptance Criteria:**')
        task.acceptanceCriteria.forEach(ac => lines.push(`- [ ] ${ac}`))
      }
      lines.push('')
    }
  }

  lines.push('## Summary Table', '')
  lines.push('| Task | Priority | Points | Requirements |')
  lines.push('|---|---|---|---|')
  for (const t of tasks) {
    lines.push(`| ${t.id}: ${t.title} | ${t.priority} | ${t.storyPoints} | ${t.requirementIds.join(', ')} |`)
  }

  return lines.join('\n')
}

export function formatSprintFeedAsGithubIssues(tasks: SprintTask[]): string {
  return tasks.map(t => JSON.stringify({
    title: `[${t.id}] ${t.title}`,
    body: [
      `**Traces:** ${t.requirementIds.join(', ')}`,
      `**Story Points:** ${t.storyPoints}`,
      '',
      '## Acceptance Criteria',
      ...t.acceptanceCriteria.map(ac => `- [ ] ${ac}`)
    ].join('\n'),
    labels: [t.priority, t.module ?? 'general'].filter(Boolean)
  })).join('\n')
}
